cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller_test.go

```diff
@@ -991,3 +992,97 @@ func TestControllerMachineSetNodeNamesUsingStatusNodeRefName(t *testing.T) {
 		}
 	}
 }
+
+func TestControllerGetAPIVersionGroup(t *testing.T) {
+	expected := "myversion.mygroup"
+	err := os.Setenv(machineAPIVersionGroupEnvVar, expected)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	observed := getAPIVersionGroup()
+	if observed != expected {
+		t.Fatalf("Wrong Version Group detected, expected %v, got %v", expected, observed)
+	}
+
+	expected = defaultMachineAPIVersionGroup
+	err = os.Setenv(machineAPIVersionGroupEnvVar, "")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	observed = getAPIVersionGroup()
+	if observed != expected {
+		t.Fatalf("Wrong Version Group detected, expected %v, got %v", expected, observed)
+	}
+}
+
+func TestControllerGetAPIVersionGroupWithMachineSets(t *testing.T) {
+	testConfig := createMachineSetTestConfig(testNamespace, 1, map[string]string{
+		nodeGroupMinSizeAnnotationKey: "1",
+		nodeGroupMaxSizeAnnotationKey: "1",
+	})
+	err := os.Setenv(machineAPIVersionGroupEnvVar, "myversion.mygroup.io")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	testConfig.machineSet.TypeMeta.APIVersion = "mygroup.io/myversion"
+	for _, machine := range testConfig.machines {
+		machine.TypeMeta.APIVersion = "mygroup.io/myversion"
+	}
+	controller, stop := mustCreateTestController(t, testConfig)
+	defer stop()
+	machineSets, err := controller.listMachineSets(testNamespace, labels.Everything())
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if l := len(machineSets); l != 1 {
+		t.Fatalf("Incorrect number of MachineSets, expected 1, got %v", l)
+	}
+	machines, err := controller.listMachines(testNamespace, labels.Everything())
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if l := len(machines); l != 1 {
+		t.Fatalf("Incorrect number of Machines, expected 1, got %v", l)
+	}
+
+	err = os.Setenv(machineAPIVersionGroupEnvVar, "")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+}
+
+func TestControllerGetAPIVersionGroupWithMachineDeployments(t *testing.T) {
+	testConfig := createMachineDeploymentTestConfig(testNamespace, 1, map[string]string{
+		nodeGroupMinSizeAnnotationKey: "1",
+		nodeGroupMaxSizeAnnotationKey: "1",
+	})
+	err := os.Setenv(machineAPIVersionGroupEnvVar, "myversion.mygroup.io")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	testConfig.machineDeployment.TypeMeta.APIVersion = "mygroup.io/myversion"
+	for _, machine := range testConfig.machines {
+		machine.TypeMeta.APIVersion = "mygroup.io/myversion"
+	}
+	controller, stop := mustCreateTestController(t, testConfig)
+	defer stop()
+	machineDeployments, err := controller.listMachineDeployments(testNamespace, labels.Everything())
```

Comment:
I wonder if this test should be checking MachineSets as well, MachineDeployments create MachineSets which create Machines right? If we are checking Machines, feels to me like we should be checking the intermediary as well?

Reply:
there is a test specifically for MachineSets, but i can add it to this one as well. i was a little confused about some of the logic around MachineSet vs MachineDeployment and i thought they were mutually exclusive.

the main thing i wanted to make sure of is that when we change the version group that the objects were found. if i add MachineSets to this test, could i just drop the specific test for them?

Reply:
So a MachineDeployment is responsible for creating MachineSets which are then responsible for creating Machines, it's a like a K8s Deployment creates replicasets creates pods

Since there is a slight semantic difference between a MachineSet owned by a Deployment and not, I would add them here but leave the explicit test too

Reply:
appreciate the explanation. i will add it, thanks!